### Задание по программированию: Выбор семейства распределений в наивном байесе

В этом задании вы сможете немного расслабиться после реализации случайного леса и градиентного бустинга по мотивам материалов прошлой недели. Все, что нужно будет делать — запускать методы из sklearn. Вам предлагается выяснить, какое распределение лучше использовать в наивном байесовском классификаторе в зависимости от вида признаков.

Загрузите датасеты digits и breast_cancer из sklearn.datasets. Выведите несколько строчек из обучающих выборок и посмотрите на признаки. С помощью sklearn.cross_validation.cross_val_score c настройками по умолчанию и вызова метода mean() у возвращаемого этой функцией numpy.ndarray, сравните качество работы наивных байесовских классификаторов на этих двух датасетах. Для сравнения предлагается использовать BernoulliNB, MultinomialNB и GaussianNB. Насколько полученные результаты согласуются с рекомендациями из лекций?

Два датасета, конечно, еще не повод делать далеко идущие выводы, но при желании вы можете продолжить исследование на других выборках (например, из UCI репозитория). 

Для сдачи задания, ответьте на приведенные ниже вопросы.

In [34]:
from sklearn import datasets, model_selection, naive_bayes

In [35]:
digits = datasets.load_digits()
breast_cancer = datasets.load_breast_cancer()

In [36]:
digits.keys()

dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

In [37]:
breast_cancer.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])

In [38]:
X_digits = digits.data
y_digits = digits.target

X_breast_cancer = breast_cancer.data
y_breast_cancer = breast_cancer.target

In [39]:
X_digits

In [40]:
X_breast_cancer

In [42]:

estimator_BernoulliNB = naive_bayes.BernoulliNB()
estimator_MultinomialNB = naive_bayes.MultinomialNB()
estimator_GaussianNB = naive_bayes.GaussianNB()

In [44]:
scoring_digits_BernoulliNB = model_selection.cross_val_score(estimator_BernoulliNB, X_digits, y_digits)
scoring_digits_BernoulliNB = scoring_digits_BernoulliNB.mean(0)

scoring_digits_MultinomialNB = model_selection.cross_val_score(estimator_MultinomialNB, X_digits, y_digits)
scoring_digits_MultinomialNB = scoring_digits_MultinomialNB.mean(0)

scoring_digits_GaussianNB = model_selection.cross_val_score(estimator_GaussianNB, X_digits, y_digits)
scoring_digits_GaussianNB = scoring_digits_GaussianNB.mean(0)

print('scoring_digits_BernoulliNB = ', scoring_digits_BernoulliNB)
print('scoring_digits_MultinomialNB = ', scoring_digits_MultinomialNB)
print('scoring_digits_GaussianNB = ', scoring_digits_GaussianNB)

scoring_digits_BernoulliNB =  0.8241736304549674
scoring_digits_MultinomialNB =  0.8703497369235531
scoring_digits_GaussianNB =  0.8069281956050759


In [51]:
scoring_breast_cancer_BernoulliNB = model_selection.cross_val_score(estimator_BernoulliNB, X_breast_cancer, y_breast_cancer)
scoring_breast_cancer_BernoulliNB = scoring_breast_cancer_BernoulliNB.mean(0)

scoring_breast_cancer_MultinomialNB = model_selection.cross_val_score(estimator_MultinomialNB, X_breast_cancer, y_breast_cancer)
scoring_breast_cancer_MultinomialNB = scoring_breast_cancer_MultinomialNB.mean(0)

scoring_breast_cancer_GaussianNB = model_selection.cross_val_score(estimator_GaussianNB, X_breast_cancer, y_breast_cancer)
scoring_breast_cancer_GaussianNB = scoring_breast_cancer_GaussianNB.mean(0)

print('scoring_breast_cancer_BernoulliNB = ', scoring_breast_cancer_BernoulliNB)
print('scoring_breast_cancer_MultinomialNB = ', scoring_breast_cancer_MultinomialNB)
print('scoring_breast_cancer_GaussianNB = ', scoring_breast_cancer_GaussianNB)

scoring_breast_cancer_BernoulliNB =  0.6274181027790716
scoring_breast_cancer_MultinomialNB =  0.8963204471355379
scoring_breast_cancer_GaussianNB =  0.9385188635305075


#### Вопрос 1

Каким получилось максимальное качество классификации на датасете breast_cancer?

In [53]:
def write_answer(answer, filename):
    with open(filename, 'w') as fout:
        fout.write(str(answer))

write_answer(scoring_breast_cancer_GaussianNB, "ans1.txt")

#### vibo: ответ на задание 1. scoring_breast_cancer_GaussianNB =  0.9385 (ответа анализатора: Верно. В датасете breast_cancer признаки вещественные, и лучшее качество получилось у GaussianNB)

#### Вопрос 2

Каким получилось максимальное качество классификации на датасете digits?

In [55]:
write_answer(scoring_digits_MultinomialNB, "ans2.txt")

#### vibo: ответ на задание 2. scoring_digits_MultinomialNB =  0.8703 (ответа анализатора: Верно. В датасете digits признаки - целые неотрицательные числа, поэтому MultinomialNB отработал лучше других вариантов.)

#### Вопрос 3

Выберите верные утверждения и запишите их номера через пробел (в порядке возрастания номера):

1) На вещественных признаках лучше всего сработал наивный байесовский классификатор с распределением Бернулли

2) На вещественных признаках лучше всего сработал наивный байесовский классификатор с мультиномиальным распределением

3) Мультиномиальное распределение лучше показало себя на выборке с целыми неотрицательными значениями признаков

4) На вещественных признаках лучше всего сработало нормальное распределение

In [56]:
write_answer('3 4', "ans3.txt")

#### vibo: ответ на задание 3. 3 и 4 верно (ответ анализатора: Мультиномиальное распределение уместней на целых неотрицательных признаках, а гауссовское - на вещественных)